File: lib/StateUI.Maui/Sources/Rendering/HandlerDispatch.cs
// SPDX-FileCopyrightText: 2026 Paweł Krzywdziński and Contributors
// SPDX-License-Identifier: Apache-2.0

namespace StateUI.Maui.Rendering;

/// <summary>
/// Raises the application's handlers in the order the reader and the platform
/// gave them - holding them while a message is being applied and raising each
/// once it is in, so no handler runs inside an apply and none is lost to one.
/// </summary>
/// <remarks>
/// <para>
/// A HANDLER RAISED INSIDE AN APPLY is Swift entered while this side is half way
/// through writing a tree Swift described: a render it asks for is computed
/// against a tree nobody finished, and what it writes lands under a message
/// still being taken. So a message holds the dispatch for as long as it is
/// being rendered - its retries included - and the handlers wait, in the order
/// they came, for its generation to be claimed.
/// </para>
/// <para>
/// What waits is the platform's and the reader's own word - a window the system
/// restored, a journey that landed, a scene's phase. A control's report of a
/// value the apply itself wrote is no word at all, and is refused before it
/// gets here: see <see cref="StateUIRenderer.Raise(object?, Protocol.HostEvent, byte[])"/>.
/// </para>
/// </remarks>
internal sealed class HandlerDispatch
{
    private readonly Action<int, byte[]?> _raise;
    private readonly Queue<(int Handler, byte[]? Payload)> _held = new();
    private int _holds;

    /// <summary>A dispatch that raises through <paramref name="raise"/>.</summary>
    /// <param name="raise">What reaches a handler: its id and its payload's wire bytes.</param>
    internal HandlerDispatch(Action<int, byte[]?> raise) => _raise = raise;

    /// <summary>Whether handlers are being held.</summary>
    internal bool Held => _holds > 0;

    /// <summary>Raises a handler, or holds it until what is being applied is in.</summary>
    /// <param name="handler">The handler's id - negative for a completion.</param>
    /// <param name="payload">The payload's wire bytes, or null for none.</param>
    internal void Raise(int handler, byte[]? payload)
    {
        if (_holds > 0)
        {
            _held.Enqueue((handler, payload));
            return;
        }

        _raise(handler, payload);
    }

    /// <summary>Holds every handler raised from now until the matching <see cref="Release"/>.</summary>
    internal void Hold() => _holds++;

    /// <summary>Lets go of one hold; the last raises what waited, in the order it came.</summary>
    internal void Release()
    {
        _holds--;

        // One at a time, and asked again after each: a handler raised here may
        // render, and a render holds and releases in its turn - which raises
        // the rest of the queue there, still in order.
        while (_holds == 0 && _held.TryDequeue(out (int Handler, byte[]? Payload) waiting))
        {
            _raise(waiting.Handler, waiting.Payload);
        }
    }
}

File: lib/StateUI/Sources/Host/Runtime/HandlerDispatch.swift
// SPDX-FileCopyrightText: 2026 Paweł Krzywdziński and Contributors
// SPDX-License-Identifier: Apache-2.0

/// The application's handlers, raised in the order they came: held while a patch applies or the user's
/// transaction runs, and raised once it is over.
/// Design: docs/design/host/runtime.md#the-handlers-order
@_spi(Host) @MainActor public final class HandlerDispatch {
    private let core: CoreLink
    private let intake: PatchIntake

    /// Handlers waiting for their turn, in the order they came.
    private var queued: [Queued] = []

    /// How deep the user's transactions stand.
    private var transactions = 0

    private struct Queued {
        let handler: Int32
        let payload: [HostValue]
        let isPhase: Bool
    }

    /// Handlers raised through `core`, held while `intake` applies a message.
    public init(core: CoreLink, intake: PatchIntake) {
        self.core = core
        self.intake = intake
    }

    /// Whether a handler raised now waits: a message applies, or the user's transaction runs.
    public var isHeld: Bool { intake.isApplying || transactions > 0 }

    /// Whether the user's transaction runs.
    public var inTransaction: Bool { transactions > 0 }

    /// Raises `handler` with `payload` now, or queues it while held; whether it ran.
    @discardableResult
    public func raise(_ handler: Int32, payload: [HostValue] = []) -> Bool {
        guard !isHeld else {
            queued.append(Queued(handler: handler, payload: payload, isPhase: false))
            return false
        }

        _ = core.dispatch(handler, payload: payload)
        return true
    }

    /// Queues a page's or a window's phase: it runs in its turn, and is rendered before anything after it.
    public func enqueuePhase(_ handler: Int32) {
        queued.append(Queued(handler: handler, payload: [], isPhase: true))
    }

    /// Runs `body` as the user's transaction; whether it was the outermost.
    func transaction(_ body: () -> Void) -> Bool {
        transactions += 1
        body()
        transactions -= 1
        return transactions == 0
    }

    /// Raises what waits, in order, stopping after a phase so it is rendered first; whether any ran.
    func raiseQueued() -> Bool {
        guard !queued.isEmpty, !isHeld else { return false }

        while !queued.isEmpty {
            let each = queued.removeFirst()
            _ = core.dispatch(each.handler, payload: each.payload)
            if each.isPhase { break }
        }
        return true
    }
}
